table.go

```diff
@@ -314,6 +345,18 @@ func (t *Table) ClearRows() {
 	t.lines = [][][]string{}
 }
 
+// ClearHeaders removes all the table's headers.
+func (t *Table) ClearHeaders() {
+	t.headers = t.headers[0:0]
+
+	// reset the column separators, otherwise we will have empty headers if the length of the new headers is smaller than the current one.
+	if len(t.cs) > 0 {
+		for k := range t.cs {
```

Comment:
```
t.cs = make(map[int]int)
```

Reply:
But, based on the next version of Go 1.11, this deletion pattern is faster than assign a new map

Reply:
Where is the source? If the map have huge elements, the loop will be slower than assigning new map. Asigning new map wll do GC in lazy.

Reply:
As far as I can see, it is optimization that replace loop for clearing map to builtin func. This doesn't mean assigning new map will be slower than clearing-map.

```go
package main_test

import (
	"runtime"
	"testing"
)

type v struct {
	m map[int]int
}

func BenchmarkLoopDelete(b *testing.B) {
	c := b.N / 50
	var v v
	v.m = make(map[int]int, c)
	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	runtime.GC()
	b.ResetTimer()
	for k := range v.m {
		delete(v.m, k)
	}
}

func BenchmarkAssignNew(b *testing.B) {
	c := b.N / 50
	var v v
	v.m = make(map[int]int, c)
	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	runtime.GC()
	b.ResetTimer()
	v.m = make(map[int]int)
}
```

```
C:\test>go test -bench . -count 1
goos: windows
goarch: amd64
BenchmarkLoopDelete-4           2000000000               0.03 ns/op
BenchmarkAssignNew-4            2000000000               0.00 ns/op
PASS
ok      _/C_/dev/go-sandbox/clear-map   14.320s
```

Reply:
Yes but in your tests you don't re-add values, this is what the optimisation really shines, in our use-case the user will re-add headers :)

Reply:
Yes but in your tests you don't re-add values, this is where the optimisation really shines, in our use-case the user will re-add headers, so make again and see the results, I already did before add that code :)

Reply:
```go
package main_test

import (
	"runtime"
	"testing"
)

type v struct {
	m map[int]int
}

func BenchmarkLoopDelete(b *testing.B) {
	c := b.N / 50
	var v v
	v.m = make(map[int]int, c)
	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	runtime.GC()
	b.ResetTimer()
	for k := range v.m {
		delete(v.m, k)
	}

	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	for k := range v.m {
		delete(v.m, k)
	}

}

func BenchmarkAssignNew(b *testing.B) {
	c := b.N / 50
	var v v
	v.m = make(map[int]int, c)
	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	runtime.GC()
	b.ResetTimer()
	v.m = make(map[int]int)

	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	v.m = make(map[int]int)
}

```
➜  mapbenchmark go test -v -bench=.
goos: linux
goarch: amd64
BenchmarkLoopDelete-8           300000000                3.89 ns/op
BenchmarkAssignNew-8            300000000                4.12 ns/op
PASS
ok      _/home/kataras/Desktop/mapbenchmark     5.402s
```

The difference it's not big and even not noticable but the loop optimisation runs a bit faster, I agree, it's not a problem assigning a new map instead of looping for deletion and vice versa, it's mostly a preference, right?

Reply:
Based on your code;

```
➜  mapbenchmark go test -v -bench=.
goos: linux
goarch: amd64
BenchmarkLoopDelete-8           300000000                3.89 ns/op
BenchmarkAssignNew-8            300000000                4.12 ns/op
PASS
ok      _/home/kataras/Desktop/mapbenchmark     5.402s
```

The difference it's not big and even not noticable but the loop optimisation runs a bit faster, I agree, it's not a problem assigning a new map instead of looping for deletion and vice versa, it's mostly a preference, right?

```go
package main_test

import (
	"runtime"
	"testing"
)

type v struct {
	m map[int]int
}

func BenchmarkLoopDelete(b *testing.B) {
	c := b.N / 50
	var v v
	v.m = make(map[int]int, c)
	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	runtime.GC()
	b.ResetTimer()
	for k := range v.m {
		delete(v.m, k)
	}

	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	for k := range v.m {
		delete(v.m, k)
	}

}

func BenchmarkAssignNew(b *testing.B) {
	c := b.N / 50
	var v v
	v.m = make(map[int]int, c)
	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	runtime.GC()
	b.ResetTimer()
	v.m = make(map[int]int)

	for i := 0; i < c; i++ {
		v.m[i] = i + 1
	}

	v.m = make(map[int]int)
}

```